### PR TITLE
[agent] Implement Backhaul STA Steering

### LINKED ISSUE #408: [TASK] Backhaul STA steering
## Description

The agent should perform steering of the backhaul STA interface based on the controller's backhaul STA steering request.

## Exit criteria

- BWL implementation of backhaul STA steering on nl80211.
- BWL implementation of backhaul STA steering on dummy.
- Agent does backhaul steering according to policy message.
- Test flow for the above.
- Manually tested on Turris Omnia.

### PR DESCRIPTION
**Pull request by VladyslavTupikin**
## Description

According to UML diagram need to do an implementation for the _**Backhaul STA Steering**_:

1. Add implementation for the **_associate()_** method in __bwl/dwpal__: it should send `NETWORK_ENABLE `message to the **_wpa suplicant_**.
2. Add implementation for the **_Backhaul STA Steering Request_** message handler in the **_backhaul_manager_**: method should parse TLV and if channel from TLV correct, trigger the **_associate()_** method.
3. Add implementation for the **_Backhaul STA Steering Response_** message creator method in the **_backhaul_manager_**.
4. Add reaction in Backhaul manager for getting `ENABLED `event from the **_wpa supplicant_** in the **_backhaul_manager_**: **_backhaul_manager_** should check specific flag which was turned on in the handler, and **if it is true**, send  **_Backhaul STA Steering Response_** message to the controller.
5. Add implementation for the **_Backhaul STA Steering Response_** message in the controller code.

https://jira.prplfoundation.org/browse/PPM-168

### CODE REVIEW

File: common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp [hunk 1]
```diff
@@ -242,7 +242,42 @@ bool sta_wlan_hal_dwpal::connect(const std::string &ssid, const std::string &pas
 
 bool sta_wlan_hal_dwpal::associate(const sMacAddr &bssid, uint8_t channel)
 {
-    LOG(DEBUG) << "sta_wlan_hal_dwpal::associate() is not implemented!";
+    auto bssid_str = tlvf::mac_to_string(bssid);
+
+    LOG(DEBUG) << " Associate iface " << get_iface_name() << " to BSSID = " << bssid_str
+               << ", Channel = " << int(channel);
```
Comment (arnout) [inline common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp]: As mentioned in #1424 this function isn't needed. It was wrong anyway because it didn't do any steering.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 2]
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
     return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
 }
 
+bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                        const std::string &src_mac)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received BACKHAUL_STA_STEERING message, mid=" << std::hex << int(mid);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
    LOG(DEBUG) << "Received BACKHAUL_STA_STEERING message, mid=" << std::hex << mid;
```
Explicit conversion to int is only needed for enums and for int8_t. It used to be needed for uint8_t as well, but that has been solved by adding an explicit operator<< overload.

And anyway, mid is uint16_t so this was never needed.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 3]
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
...
+
+    auto bh_sta_steering_req = cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>();
+    if (!bh_sta_steering_req) {
+        LOG(ERROR) << "Failed cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>(), mid="
+                   << std::hex << int(mid);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
                   << std::hex << mid;
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 4]
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
...
 
+bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                        const std::string &src_mac)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received BACKHAUL_STA_STEERING message, mid=" << std::hex << int(mid);
+
+    auto bh_sta_steering_req = cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>();
+    if (!bh_sta_steering_req) {
+        LOG(ERROR) << "Failed cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>(), mid="
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
        LOG(WARNING) << "Failed cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>(), mid="
```
Since this message comes from outside, it is not an "error" (in the sense that something is wrong with our system). Rather , it is invalid input.

Doesn't matter much of course, but we should probably do this right.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 5]
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
...
+              Response with the error code without trigerring HAL.
+    */
+
+    auto active_hal = get_wireless_hal();
+    if (!active_hal) {
+        LOG(ERROR) << "Couldn't get active HAL";
+        return false;
+    }
+
+    auto associate = active_hal->associate(bssid, channel);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
    auto associate = active_hal->roam(bssid, channel);
```

The proper way is probably to use the state machine instead. That said, maybe it's not needed, and it's way more efficient if we can roam directly here rather than going through the state machine.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 6]
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
...
+        LOG(ERROR) << "Couldn't associate active HAL with bssid: " << tlvf::mac_to_string(bssid);
+    }
+
+    // build ACK message CMDU
+    auto cmdu_tx_header = cmdu_tx.create(mid, ieee1905_1::eMessageType::ACK_MESSAGE);
+    if (!cmdu_tx_header) {
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+    LOG(DEBUG) << "sending ACK message to the originator, mid=" << std::hex << int(mid);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
    LOG(DEBUG) << "sending ACK message to the originator, mid=" << std::hex << mid;
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 7]
```diff
@@ -4507,6 +4506,31 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
     return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
 }
 
+bool backhaul_manager::create_backhaul_steering_response(
+    const sMacAddr &sta_mac, const sMacAddr &target_bssid,
+    const wfa_map::tlvBackhaulSteeringResponse::eResultCode &result_code)
+{
+    auto cmdu_tx_header = m_expected_backhaul_sta_steering.response.create(
+        0, ieee1905_1::eMessageType::BACKHAUL_STEERING_RESPONSE_MESSAGE);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Since we are not building up the message from different responses here but we're just expecting a single one, there's no need to keep a response buffer in m_expected_backhaul_sta_steering. We can simply use cmdu_tx here.
```suggestion
    auto cmdu_tx_header = cmdu_tx.create(
        0, ieee1905_1::eMessageType::BACKHAUL_STEERING_RESPONSE_MESSAGE);
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 8]
```diff
@@ -886,6 +886,14 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
                 FSM_MOVE_STATE(MASTER_DISCOVERY);
             }
 
+            if (m_expected_backhaul_sta_steering.backhaul_sta_steering_enable) {
+                LOG(DEBUG) << "Sending BACKHAUL_STA_STEERING_RESPONSE_MESSAGE";
+
+                m_expected_backhaul_sta_steering.backhaul_sta_steering_enable = false;
+
+                send_cmdu_to_bus(m_expected_backhaul_sta_steering.response, controller_bridge_mac,
+                                 bridge_info.mac);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Here, we should also create the response.

The bssid is active_hal->get_bssid(). Funnily enough, the station MAC address is not so easy to get: you have to get it from active_hal->get_iface_name() using get_iface_mac().

File: controller/src/beerocks/master/son_master_thread.cpp [hunk 9]
```diff
@@ -1794,6 +1799,25 @@ bool master_thread::handle_cmdu_1905_topology_response(const std::string &src_ma
     return true;
 }
 
+bool master_thread::handle_cmdu_1905_backhaul_sta_steering_response(
+    const std::string &src_mac, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received BACKHAUL_STA_STEERING_MESSAGE from " << src_mac << ", mid=" << std::hex
+               << int(mid);
```
Comment (arnout) [inline controller/src/beerocks/master/son_master_thread.cpp]:
```suggestion
               << mid;
```

**Review by arnout — APPROVED** [review]
Although some comments are pretty major, since we don't have any way to test this yet there's no point gating the merge on that. So accepting.

File: common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp (hunk 1 above)
Comment (VladyslavTupikin) [reply common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp]: This commit removed.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 2 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 3 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 4 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: accepted!

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 5 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Roger that.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 6 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 7 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: accepted

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 8 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: solved

File: controller/src/beerocks/master/son_master_thread.cpp (hunk 9 above)
Comment (VladyslavTupikin) [reply controller/src/beerocks/master/son_master_thread.cpp]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 10]
```diff
@@ -4580,6 +4579,41 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
     return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
 }
 
+bool backhaul_manager::create_backhaul_steering_response(
+    const wfa_map::tlvBackhaulSteeringResponse::eResultCode &result_code)
+{
+    auto cmdu_tx_header =
+        cmdu_tx.create(0, ieee1905_1::eMessageType::BACKHAUL_STEERING_RESPONSE_MESSAGE);
```
Comment (vitalii-komisarenko) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:4625]: Is it OK to use message ID 0?
Often message ID of the request and the response should be the same.
Certification test 4.9.1 does not check it. Can it change in the future?

**Review by vitalii-komisarenko — APPROVED** [review]
Some commit messages are too long.
"target_bssid" is menioned two times in the commit message of 7454c656c19467af759282066cc7f793f247f056.
Other than that looks OK.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 11]
```diff
@@ -4505,6 +4511,58 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
...
+    /*
+        TODO: Need to do channel validation. If channel invalid, send ACK and Backhaul STA Steering
+              Response with the error code without trigerring HAL.
+    */
+
+    auto active_hal = get_wireless_hal();
+    if (!active_hal) {
+        LOG(ERROR) << "Couldn't get active HAL";
+        return false;
+    }
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:4591]: This can actually happen if we have wired backhaul. I'm not sure what to do in that case, but it would be good to at least add a TODO here. I think we should still send an ACK anyway. Sending ACK should be done before getting active hal.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 12]
```diff
@@ -4505,6 +4511,58 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
...
+        /*
+            TODO: Need to add sending Backhaul STA Steering Response with the FAILURE error code.
+        */
+
+        return false;
+    }
+
+    LOG(DEBUG) << "Sending ACK message to the originator, mid=" << std::hex << mid;
+    return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
+}
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Sending ACK should already be done before doing the roam, because we have to do it even if the roam fails.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 13]
```diff
@@ -4563,6 +4562,41 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
     return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
 }
 
+bool backhaul_manager::create_backhaul_steering_response(
+    const wfa_map::tlvBackhaulSteeringResponse::eResultCode &result_code)
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I think we should have as parameter the error code rather than the result code (see below).

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 10 above)
Comment (arnout) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:4625]: Yes it has to be a new MID. The request is answered with an ACK that has the same MID as the request, while the response uses a new MID. It's a bit confusing...

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 14]
```diff
@@ -4563,6 +4562,41 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
...
+    if (!active_hal) {
+        LOG(ERROR) << "Couldn't get active HAL";
+        return false;
+    }
+
+    sMacAddr sta_mac;
+    auto interface = active_hal->get_iface_name();
+    get_iface_mac(interface, sta_mac);
+
+    LOG(DEBUG) << "Interface: " << interface << "MAC: " << tlvf::mac_to_string(sta_mac);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
    LOG(DEBUG) << "Interface: " << interface << "MAC: " << sta_mac;
```

**Review by vitalii-komisarenko — APPROVED** [review]
My review is identical to previous one: https://github.com/prplfoundation/prplMesh/pull/1445#pullrequestreview-438281483
Mostly related to commit messages.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 15]
```diff
@@ -4563,6 +4562,41 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
...
+
+    bh_steering_resp_tlv->target_bssid()         = tlvf::mac_from_string(active_hal->get_bssid());
+    bh_steering_resp_tlv->backhaul_station_mac() = sta_mac;
+    bh_steering_resp_tlv->result_code()          = result_code;
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: If result code is failure, we also have to provide an error code TLV with the error code. See the 4th paragraph of section 12.

So I think it's better to pass the error code directly, and to set the result code (and include the error TLV) depending on whether the error code is 0 or not.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h [hunk 16]
```diff
@@ -597,6 +601,8 @@ class backhaul_manager : public btl::transport_socket_thread {
 
     sExpectedChannelSelection m_expected_channel_selection;
 
+    bool backhaul_sta_steering_enable = false;
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: This change is unrelated and not needed I think.

**Review by arnout — COMMENTED** [review]
Only reviewed the first 2 commits, but I have to take a break now :-)

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 17]
```diff
@@ -969,6 +969,16 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
         eth_link_poll_timer = std::chrono::steady_clock::now();
         m_eth_link_up       = network_utils::linux_iface_is_up_and_running(m_sConfig.wire_iface);
         FSM_MOVE_STATE(OPERATIONAL);
+
+        if (backhaul_sta_steering_enable) {
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: This needs a comment IMO. Something like:

```
// This event may come as a result of enabling the backhaul, but also as a result
// of steering. *Only* in case it was the result of steering, we need to send a steering
// response.
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 18]
```diff
@@ -4551,13 +4561,21 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
         LOG(DEBUG) << "Sending ACK message to the originator, mid=" << std::hex << mid;
         send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
 
-        /*
-            TODO: Need to add sending Backhaul STA Steering Response with the FAILURE error code.
-        */
+        auto response = create_backhaul_steering_response(
+            wfa_map::tlvBackhaulSteeringResponse::eResultCode::FAILURE);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: As mentioned earlier, in case of failure we also need to give an error code. We have the choice of BACKHAUL_STEERING_REQUEST_REJECTED_CANNOT_OPERATE_ON_CHANNEL_SPECIFIED, BACKHAUL_STEERING_REQUEST_REJECTED_TARGET_BSS_SIGNAL_NOT_SUITABLE or BACKHAUL_STEERING_REQUEST_AUTHENTICATION_OR_ASSOCIATION_REJECTED. Actually none of these are applicable here because if the roam() call failed, it just means there's a problem with the connection to wpa_supplicant. I think BACKHAUL_STEERING_REQUEST_REJECTED_TARGET_BSS_SIGNAL_NOT_SUITABLE is the closest to be accurate.

File: controller/src/beerocks/master/son_master_thread.cpp [hunk 19]
```diff
@@ -1794,6 +1799,25 @@ bool master_thread::handle_cmdu_1905_topology_response(const std::string &src_ma
...
+
+    auto tlv_backhaul_sta_steering_resp = cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringResponse>();
+    if (!tlv_backhaul_sta_steering_resp) {
+        LOG(ERROR) << "Failed getClass<wfa_map::tlvBackhaulSteeringResponse>";
+        return false;
+    }
+
+    LOG(DEBUG) << "BACKHAUL_STA_STEERING_MESSAGE result_code: "
+               << int(tlv_backhaul_sta_steering_resp->result_code());
+
```
Comment (arnout) [inline controller/src/beerocks/master/son_master_thread.cpp:1817]: The controller should respond with an ACK.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 20]
```diff
@@ -4554,6 +4544,33 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
...
+    auto is_valid_channel = son::wireless_utils::is_channel_in_operating_class(oper_class, channel);
+
+    if (!is_valid_channel) {
+        LOG(WARNING) << "Invalid channel number";
+
+        LOG(DEBUG) << "Sending ACK message to the originator, mid=" << std::hex << mid;
+        send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
+
+        auto response = create_backhaul_steering_response(
+            wfa_map::tlvBackhaulSteeringResponse::eResultCode::FAILURE);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Here the error code must be BACKHAUL_STEERING_REQUEST_REJECTED_CANNOT_OPERATE_ON_CHANNEL_SPECIFIED.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 21]
```diff
@@ -4554,6 +4544,33 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
         return false;
     }
 
+    auto is_valid_channel = son::wireless_utils::is_channel_in_operating_class(oper_class, channel);
+
+    if (!is_valid_channel) {
+        LOG(WARNING) << "Invalid channel number";
+
+        LOG(DEBUG) << "Sending ACK message to the originator, mid=" << std::hex << mid;
+        send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: This is independent on whether the channel is valid or not, so it can be done before is_valid_channel.

**Review by arnout — APPROVED** [review]
Except for including an error code TLV in the response in case of failure, none of my comments are really critical. Also there's no need for me to re-review.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 11 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:4591]: Added TODO above getting HAL. Sending ACK message moved upper than getting active HAL.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 13 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: accepted!

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 14 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 15 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: You are right. Added creation for ErrorCodeTlv

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 12 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Accepted

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h (hunk 16 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Moved to commit with usage.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 17 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Accepted

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 18 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Accepted.

File: controller/src/beerocks/master/son_master_thread.cpp (hunk 19 above)
Comment (VladyslavTupikin) [reply controller/src/beerocks/master/son_master_thread.cpp:1817]: Added sending ACK message to the agent and fixed commit message.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 20 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 21 above)
Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Accepted

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 22]
```diff
@@ -4517,6 +4535,139 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
...
+            wfa_map::tlvBackhaulSteeringResponse::eResultCode::FAILURE;
+    }
+
+    auto error_code_tlv = cmdu_tx.addClass<wfa_map::tlvErrorCode>();
+    if (!bh_steering_resp_tlv) {
+        LOG(ERROR) << "Couldn't addClass<wfa_map::tlvErrorCode>";
+        return false;
+    }
+
+    error_code_tlv->reason_code() = error_code;
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: This should only be added if error_code != 0.

Comment (VladyslavTupikin) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: fixed
